src/masternodes/mn_checks.cpp

```diff
@@ -1616,6 +1647,68 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
         return !res ? res : mnview.ICXCloseMakeOfferTx(*offer, CICXMakeOffer::STATUS_CLOSED);
     }
 
+    Res operator()(const CCreatePropMessage& obj) const {
+        auto res = CheckProposalTx(obj.type);
+        if (!res) {
+            return res;
+        }
+        if (obj.nAmount >= MAX_MONEY) {
+            return Res::Err("proposal wants to gain all money");
+        }
+        if (obj.title.size() > 128) {
+            return Res::Err("proposal title cannot be more than 128 bytes");
+        }
+        if (obj.type == CPropType::CommunityFundRequest) {
+            if (obj.nCycles < 1 || obj.nCycles > MAX_CYCLES) {
+                return Res::Err("proposal cycles can be between 1 and %d", int(MAX_CYCLES));
+            }
+        } else if (obj.type == CPropType::VoteOfConfidence) {
+            if (obj.nAmount != 0) {
+                return Res::Err("proposal amount in vote of confidence");
+            }
+            if (obj.nCycles != VOC_CYCLES) {
+                return Res::Err("proposal cycles should be %d", int(VOC_CYCLES));
+            }
+        } else {
+            return Res::Err("unsupported proposal type");
+        }
+        return mnview.CreateProp(tx.GetHash(), height, obj, consensus.props.votingPeriod);
+    }
+
+    Res operator()(const CPropVoteMessage& obj) const {
+        auto prop = mnview.GetProp(obj.propId);
+        if (!prop) {
+            return Res::Err("proposal <%s> does not exists", obj.propId.GetHex());
+        }
+        if (prop->status != CPropStatusType::Voting) {
+            return Res::Err("proposal <%s> is not in voting period", obj.propId.GetHex());
+        }
+        auto node = mnview.GetMasternode(obj.masternodeId);
+        if (!node) {
+            return Res::Err("masternode <%s> does not exist", obj.masternodeId.GetHex());
+        }
+        auto ownerDest = node->ownerType == 1 ? CTxDestination(PKHash(node->ownerAuthAddress))
+                                              : CTxDestination(WitnessV0KeyHash(node->ownerAuthAddress));
+        if (!HasAuth(GetScriptForDestination(ownerDest))) {
+            return Res::Err("tx must have at least one input from the owner");
+        }
+        if (!node->IsActive(height)) {
+            return Res::Err("masternode <%s> is not active", obj.masternodeId.GetHex());
+        }
+        if (node->mintedBlocks < 1) {
```

Comment:
From the pink paper
`The masternode has successfully minted at least 1 block during vote submission AND at vote finalizing block.`
not sure if we have the data to check the duration part of the above statement. pink paper should be updated??

Reply:
From what I can see on `ain` side, the check is only this one. It is not following the pink paper implementation.
I will raise the issue on the governance PR.

Reply:
hmm, looks like I read it wrong. there's an`at`. so it's two conditions at vote submission and at vote finalizing.

Reply:
A check at vote finalizing block should be added then ?

Reply:
It has https://github.com/DeFiCh/ain/pull/576/commits/d3ea5d201ad52d2cd2d7778d483a1696bea1fbe5#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2878
It gets all active nodes to finalizing height then count `yes` votes when masternode is part of active ones.

Reply:
Ok sweet